subxt/src/error.rs

```diff
@@ -102,10 +102,17 @@ impl From<DispatchError> for Error {
 }
 
 /// An RPC error. Since we are generic over the RPC client that is used,
-/// the error is any custom string.
+/// the error is boxed and could be casted.
 #[derive(Debug, thiserror::Error)]
 #[error("RPC error: {0}")]
-pub struct RpcError(pub String);
+pub enum RpcError {
+    // Dev note: We need the error to be safely sent between threads
+    // for `subscribe_to_block_headers_filling_in_gaps` and friends.
+    /// Error related to the RPC client.
+    ClientError(Box<dyn std::error::Error + Send + 'static>),
+    /// Custom error.
+    Custom(String),
+}
```

Comment:
I'm thinking to myself; can we avoid having the generic "custom" variant?

Looking at the code, we have two `RpcError::Custom("RPC subscription dropped".into())` calls to deal with and "RPC Params are expected to be an array but got {params}" error to handle.

I'm also thinking; if a custom implementation needs unique error variants, it's probably a good idea if that implementation defines it's own enum with whatever variants it cares about, and we return that directly.

Sooo, how about we:

- Replace `Custom` for `SubscriptionDropped` (and use that for those dropped errors)
- Turn the jsonrpsee "RPC Params are expected to be an array but got {params}" error into a panic, and make sure it's mentioned in the `RpcClientT` trait methods that implementations are free to panic if parameters are not a JSON array. (it should be that we only use this internally anyway, and we do enforce this in the higher level interface, so I'm ok to think of it as a programmer error if this condition is not upheld).

Reply:
That does make sense! Thanks for the suggestions!

I think we should remove entirely the `prep_params_for_jsonrpsee` with the next jsonrpsee release that should include [this optimization](https://github.com/paritytech/jsonrpsee/issues/862) 😄

Reply:
Yup, that was on my mind too; we'll be able to simplify this stuff soon anyway :)